newlib/libc/tinystdio/fmemopen.c

```diff
@@ -33,119 +33,166 @@
+#define __MALL 0x01
 
 struct __file_mem {
-        struct __file_ext xfile;
-        char    *buf;
-        size_t  size;
-        size_t  pos;
-        uint8_t mflags;
+    struct __file_ext xfile;
+    char *buf;
+    size_t size; /* Current size. */
+    size_t bufsize; /* Upper limit on size. */
+    size_t pos;
+    uint8_t mflags;
 };
 
-static int __fmem_put(char c, FILE *f)
+static int
+__fmem_put(char c, FILE *f)
 {
-        struct __file_mem *mf = (struct __file_mem *) f;
-        if ((f->flags & __SWR) && mf->pos < mf->size) {
-                mf->buf[mf->pos++] = c;
-                return (unsigned char) c;
-        }
+    struct __file_mem *mf = (struct __file_mem *)f;
+    if ((f->flags & __SWR) == 0) {
         return _FDEV_ERR;
+    } else if (mf->pos < mf->bufsize) {
+        mf->buf[mf->pos++] = c;
+        if (mf->pos > mf->size) {
+            mf->size = mf->pos;
+        }
+        return (unsigned char)c;
+    } else {
+        return _FDEV_EOF;
+    }
 }
 
-static int __fmem_get(FILE *f)
+static int
+__fmem_get(FILE *f)
 {
-        struct __file_mem *mf = (struct __file_mem *) f;
-        int c;
-        if ((f->flags & __SRD) && mf->pos < mf->size) {
-                c = (unsigned char) mf->buf[mf->pos++];
-                if (c == '\0')
-                        c = _FDEV_EOF;
-        } else
-                c = _FDEV_ERR;
-        return c;
+    struct __file_mem *mf = (struct __file_mem *)f;
+    if ((f->flags & __SRD) == 0) {
+        return _FDEV_ERR;
+    } else if (mf->pos < mf->bufsize) {
+        return (unsigned char)mf->buf[mf->pos++];
+    } else {
+        return _FDEV_EOF;
+    }
 }
 
-static int __fmem_flush(FILE *f)
+static int
+__fmem_flush(FILE *f)
 {
-        struct __file_mem *mf = (struct __file_mem *) f;
-        if ((f->flags & __SWR) && mf->pos < mf->size)
-                mf->buf[mf->pos] = '\0';
-        return 0;
+    struct __file_mem *mf = (struct __file_mem *)f;
+    if ((f->flags & __SWR) && mf->pos < mf->bufsize) {
+        mf->buf[mf->pos] = '\0';
+        if (mf->pos > mf->size) {
+            mf->size = mf->pos;
+        }
+    }
+    return 0;
 }
 
-static off_t __fmem_seek(FILE *f, off_t pos, int whence)
+static off_t
+__fmem_seek(FILE *f, off_t pos, int whence)
 {
-        struct __file_mem *mf = (struct __file_mem *) f;
-
-        switch (whence) {
-        case SEEK_SET:
-                break;
-        case SEEK_CUR:
-                pos += mf->pos;
-                break;
-        case SEEK_END:
-                pos += mf->size;
-                break;
-        }
-        if (pos < 0 || mf->size < (size_t) pos)
-                return EOF;
-        mf->pos = pos;
-        return pos;
+    struct __file_mem *mf = (struct __file_mem *)f;
+
+    switch (whence) {
+    case SEEK_SET:
+        break;
+    case SEEK_CUR:
+        pos += mf->pos;
+        break;
+    case SEEK_END:
+        pos += mf->size;
+        break;
+    }
+    if (pos < 0 || mf->bufsize < (size_t)pos)
```

Comment:
Hrm. I think we need to be careful here with types -- if 'off_t' is larger than 'size_t', then you could pass a large
'pos' which truncated to an in-range value. It's "probably" safe to assume `sizeof(off_t) >= sizeof(size_t)`?